src/ir_Bosch.cpp

```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
 /// @param[in] nbytes The number of bytes of message to be sent.
 /// @param[in] repeat The number of times the command is to be repeated.
 void IRsend::sendBosch144(const unsigned char data[], const uint16_t nbytes,
                           const uint16_t repeat) {
-  // nbytes is required to be a multiple of kBosch144NrOfSections.
-  if (nbytes % kBosch144NrOfSections != 0) return;
-
+  // nbytes is required to be a multiple of kBosch144BytesPerSection.
+  if (nbytes % kBosch144BytesPerSection != 0)return;
   // Set IR carrier frequency
   enableIROut(kBoschFreq);
 
   for (uint16_t r = 0; r <= repeat; r++) {
-    const uint16_t kSectionByteSize = nbytes / kBosch144NrOfSections;
-    for (uint16_t offset = 0; offset < nbytes; offset += kSectionByteSize)
+    for (uint16_t offset=0; offset < nbytes; offset += kBosch144BytesPerSection)
       // Section Header + Data + Footer
       sendGeneric(kBoschHdrMark, kBoschHdrSpace,
                   kBoschBitMark, kBoschOneSpace,
                   kBoschBitMark, kBoschZeroSpace,
                   kBoschBitMark, kBoschFooterSpace,
-                  data + offset, kSectionByteSize,
+                  data + offset, kBosch144BytesPerSection,
                   kBoschFreq, true, 0, kDutyDefault);
     space(kDefaultMessageGap);  // Complete guess
   }
 }
 
 #endif  // SEND_BOSCH144
 
+/// Class constructor.
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRBosch144AC::IRBosch144AC(const uint16_t pin, const bool inverted,
+                           const bool use_modulation)
+    : _irsend(pin, inverted, use_modulation) { stateReset(); }
+
+/// Reset the internal state to a fixed known good state.
+void IRBosch144AC::stateReset(void) {
+  setRaw(kBosch144DefaultState);
+  setPower(true);
+}
+
+/// Set up hardware to be able to send a message.
+void IRBosch144AC::begin(void) { _irsend.begin(); }
+
+#if SEND_BOSCH144
+/// Send the current internal state as an IR message.
+/// @param[in] repeat Nr. of times the message will be repeated.
+void IRBosch144AC::send(const uint16_t repeat) {
+  if (!powerFlag) {  // "Off" is a 96bit message
+    _irsend.sendBosch144(kBosch144Off, sizeof(kBosch144Off), repeat);
+  } else {
+    _irsend.sendBosch144(getRaw(), kBosch144StateLength, repeat);
+  }   // other 96bit messages are not yet supported
+}
+#endif  // SEND_BOSCH144
+
+/// Get a copy of the internal state as a valid code for this protocol.
+/// @return A valid code for this protocol based on the current internal state.
+uint8_t* IRBosch144AC::getRaw(void) {
+  setInvertBytes();
+  setCheckSumS3();
+  return _.raw;
+}
+
+/// Set the internal state from a valid code for this protocol.
+/// @param[in] new_code A valid code for this protocol.
+void IRBosch144AC::setRaw(const uint8_t new_code[]) {
+  std::memcpy(_.raw, new_code, kBosch144StateLength);
+}
```

Comment:
What happens if someone sets the "Off" message as the raw state?
You need to handle that length, and to set the appropriate `powerFlag` value.

Reply:
Could this work?
```suggestion
void IRBosch144AC::setRaw(const uint8_t new_code[]) {
  if (sizeof(new_code) == kBosch144StateLength) {
    setPower(true);
    std::memcpy(_.raw, new_code, kBosch144StateLength);
  } else {
    if (new_code == kBosch144Off) setPower(false);
  }
}
```

Reply:
it doesn't work

Reply:
`void IRBosch144AC::setRaw(const uint8_t new_code[]) {
  if (new_code == kBosch144Off) {
    setPower(false);
  } else {
    std::memcpy(_.raw, new_code, kBosch144StateLength);
  }
}`
should work